### ISSUE TITLE
KubernetesProvider not working with PartitionIdentityLookup

### ISSUE THREAD

**Issue by sudsy** [AsynkronIT/protoactor-dotnet#819]
## Expected Behavior
I have a cluster which I have recently updated to 0.14. It is working with the consul provider but it fails when I use the KubernetesProvider

## Actual Behavior
It fails with a message "Identity belongs to " . 

## Steps to Reproduce the Problem

  1. Switch from Consul provider to kubernetes provider with the following configuration

```
var kubernetes = new Kubernetes(KubernetesClientConfiguration.InClusterConfig());
var k8sProvider = new KubernetesProvider(kubernetes);
ClusterConfig.Setup("randomID" , k8sconsulProvider, new PartitionIdentityLookup());
```

## StackTrace
<details>

When I run the cluster using the Consul provider I receive the following startup messages:

```
{"Timestamp":"2021-03-26T07:12:48.3757922+00:00","Level":"Information","MessageTemplate":"Starting","Properties":{"SourceContext":"Cluster-192.168.108.208:44300"}}
{"Timestamp":"2021-03-26T07:12:48.4731207+00:00","Level":"Information","MessageTemplate":"Started Cluster Heartbeats","Properties":{"SourceContext":"ClusterHeartBeat-192.168.108.208:44300"}}
{"Timestamp":"2021-03-26T07:12:48.9880278+00:00","Level":"Information","MessageTemplate":"Started as cluster member","Properties":{"SourceContext":"Cluster-192.168.108.208:44300"}}
{"Timestamp":"2021-03-26T07:12:48.9885420+00:00","Level":"Information","MessageTemplate":"Proto.Server Started in {HostingEnvironment}","Properties":{"HostingEnvironment":"kubernetesprod","SourceContext":"PVL.UserSettings.Svc.ProtoServerService"}}
{"Timestamp":"2021-03-26T07:12:49.3695088+00:00","Level":"Information","MessageTemplate":"Topology change --- {EventId} --- pausing interactions for {Timeout}","Properties":{"EventId":1788388,"Timeout":"00:00:03","SourceContext":"PartitionIdentityActor-192.168.108.208:44300"}}
{"Timestamp":"2021-03-26T07:12:49.4705261+00:00","Level":"Information","MessageTemplate":"Cluster members joined {MembersJoined}","Properties":{"MembersJoined":["{ \"host\": \"192.168.108.208\", \"port\": 44300, \"id\": \"674075056beb4c98828eddec4e1e1b55\" }"],"SourceContext":"MemberList-192.168.108.208:44300"}}
{"Timestamp":"2021-03-26T07:12:49.4750511+00:00","Level":"Information","MessageTemplate":"Topology change --- {EventId} --- pausing interactions for {Timeout}","Properties":{"EventId":1788389,"Timeout":"00:00:03","SourceContext":"PartitionIdentityActor-192.168.108.208:44300"}}
```

When I run using the Kubernetes provider I get the following

```
{"Timestamp":"2021-03-26T07:16:45.9708308+00:00","Level":"Information","MessageTemplate":"Starting","Properties":{"SourceContext":"Cluster-192.168.105.111:44300"}}
{"Timestamp":"2021-03-26T07:16:46.0677889+00:00","Level":"Information","MessageTemplate":"Started Cluster Heartbeats","Properties":{"SourceContext":"ClusterHeartBeat-192.168.105.111:44300"}}
{"Timestamp":"2021-03-26T07:16:46.0802476+00:00","Level":"Information","MessageTemplate":"Registering service {PodName} on {PodIp}","Properties":{"PodName":"usersettings-stage-8ff976f47-tjhnq","PodIp":"192.168.105.111:44300","SourceContext":"Proto.Cluster.Kubernetes.KubernetesProvider"}}
{"Timestamp":"2021-03-26T07:16:47.1833734+00:00","Level":"Information","MessageTemplate":"Using Kubernetes namespace: stage","Properties":{"SourceContext":"Proto.Cluster.Kubernetes.KubernetesProvider"}}
{"Timestamp":"2021-03-26T07:16:47.1840856+00:00","Level":"Information","MessageTemplate":"Using Kubernetes port: 44300","Properties":{"SourceContext":"Proto.Cluster.Kubernetes.KubernetesProvider"}}
{"Timestamp":"2021-03-26T07:16:47.4832274+00:00","Level":"Information","MessageTemplate":"Started as cluster member","Properties":{"SourceContext":"Cluster-192.168.105.111:44300"}}
{"Timestamp":"2021-03-26T07:16:47.4838869+00:00","Level":"Information","MessageTemplate":"Proto.Server Started in {HostingEnvironment}","Properties":{"HostingEnvironment":"kubernetesprod","SourceContext":"PVL.UserSettings.Svc.ProtoServerService"}}
{"Timestamp":"2021-03-26T07:16:48.1702528+00:00","Level":"Information","MessageTemplate":"Cluster members joined {MembersJoined}","Properties":{"MembersJoined":["{ \"host\": \"192.168.105.111\", \"port\": 44300, \"id\": \"1f1e6c1100144837a88f90871f125429\" }"],"SourceContext":"MemberList-192.168.105.111:44300"}}
```

I can see that the line "Topology change --- " is missing when I run the kubernetes provider and when I run the debug logging I get the message `Identity belongs to` instead of showing which cluster member the identity really belongs to.

I tried to infer from the source code which permissions would be required in kubernetes and ended up with :

```
- apiGroups: [""]
    resources: ["pods"]
    verbs: ["get", "watch", "patch"]
```

Perhaps I have missed a key permission?

</details>

**Comment by rogeralsing** [comment]
Thanks for reporting. 
I know what the issue is here.

PartitionIdentityLookup relies on cluster-wide EventIndex to know what version of a topology event it is currently using.
The Kubernetes cluster provider does not have such EventIndex. 

The recent introduction of Member.GetMembershipHashCode() can likely solve this. giving us a unique id for each topology, regardless of what ClusterProvider is used.
I have done some attempts on fixing this but did not get the results I wanted.

TLDR; we know the reason, we know the fix, we just need to glue it all together.

**Comment by sudsy** [comment]
Thanks for updating the code base,  I just updated my code to version 0.17 and things seem to have improved.

I now get the topology events when using the k8s provider as detailed below:

`"Timestamp":"2021-05-05T01:36:02.1534422+00:00","Level":"Debug","MessageTemplate":"Updating Cluster Topology","Properties":{"SourceContext":"MemberList-192.168.167.61:44300"}}
{"Timestamp":"2021-05-05T01:36:02.2623125+00:00","Level":"Debug","MessageTemplate":"Published ClusterTopology event {ClusterTopology}","Properties":{"ClusterTopology":"{ \"members\": [ { \"host\": \"192.168.167.61\", \"port\": 44300, \"id\": \"69ab68f2fa4c49269a3a5029efff64ac\", \"kinds\": [ \"calculatorsettings\", \"defects\", \"usertemplates\", \"weather\" ] } ], \"joined\": [ { \"host\": \"192.168.167.61\", \"port\": 44300, \"id\": \"69ab68f2fa4c49269a3a5029efff64ac\" } ] }","SourceContext":"MemberList-192.168.167.61:44300"}}
{"Timestamp":"2021-05-05T01:36:02.2638472+00:00","Level":"Information","MessageTemplate":"Cluster members joined {MembersJoined}","Properties":{"MembersJoined":["{ \"host\": \"192.168.167.61\", \"port\": 44300, \"id\": \"69ab68f2fa4c49269a3a5029efff64ac\" }"],"SourceContext":"MemberList-192.168.167.61:44300"}}
{"Timestamp":"2021-05-05T01:36:03.2546524+00:00","Level":"Debug","MessageTemplate":"Starting activations","Properties":{"SourceContext":"PartitionIdentityActor-192.168.167.61:44300"}}
{"Timestamp":"2021-05-05T01:36:05.2600602+00:00","Level":"Debug","MessageTemplate":"Statistics: Actor Count {ActorCount}","Properties":{"ActorCount":0,"SourceContext":"PartitionPlacementActor-192.168.167.61:44300"}}`

But when I make a call to get an actor from the cluster `var settingsActor = await _cluster.GetAsync(actorName, _actorKind);`, the cluster returns null when using the k8s cluster provider.

I can see that it has identified which cluster member owns the identity:

`{"Timestamp":"2021-05-05T01:36:09.9440544+00:00","Level":"Debug","MessageTemplate":"Identity belongs to {address}","Properties":{"address":"","SourceContext":"PartitionIdentityLookup-192.168.167.61:44300","ActionId":"4d703716-5cc8-40e3-85a9-a3f13da98b18","ActionName":"PVL.UserSettings.Svc.Controllers.PubCalculatorSettingsController.Get (PVL.UserSettings.Svc)","RequestId":"0HM8FCOJNPPFC:00000001","RequestPath":"/svc/usersettings/pubcalculatorsettings","SpanId":"|b444fcfd-40d915995a47d100.","TraceId":"b444fcfd-40d915995a47d100","ParentId":"","ConnectionId":"0HM8FCOJNPPFC"}}`

but it still returns a null PID.

When I use the consul cluster provider, the messages directly prior to receiving the actor PID look like htis:

`{"Timestamp":"2021-05-05T01:46:23.8472627+00:00","Level":"Debug","MessageTemplate":"Identity belongs to {address}","Properties":{"address":"192.168.108.208:44300","SourceContext":"PartitionIdentityLookup-192.168.108.208:44300","ActionId":"583707b7-28d4-4cb6-8325-91b582d0dd88","ActionName":"PVL.UserSettings.Svc.Controllers.PubCalculatorSettingsController.Get (PVL.UserSettings.Svc)","RequestId":"0HM7JQPP7L2SV:00000001","RequestPath":"/svc/usersettings/pubcalculatorsettings","SpanId":"|d876fcf-4b964af8013a80e2.","TraceId":"d876fcf-4b964af8013a80e2","ParentId":"","ConnectionId":"0HM7JQPP7L2SV"}}
{"Timestamp":"2021-05-05T01:46:23.8478242+00:00","Level":"Debug","MessageTemplate":"Requesting remote PID from {Partition}:{Remote} {@Request}","Properties":{"Partition":"192.168.108.208:44300","Remote":"192.168.108.208:44300/partition-identity","Request":{"_typeTag":"ActivationRequest","Kind":"calculatorsettings","Identity":"calculatorsettings/pubsettings","ClusterIdentity":{"_typeTag":"ClusterIdentity","Identity":"calculatorsettings/pubsettings","Kind":"calculatorsettings"},"RequestId":""},"SourceContext":"PartitionIdentityLookup-192.168.108.208:44300","ActionId":"583707b7-28d4-4cb6-8325-91b582d0dd88","ActionName":"PVL.UserSettings.Svc.Controllers.PubCalculatorSettingsController.Get (PVL.UserSettings.Svc)","RequestId":"0HM7JQPP7L2SV:00000001","RequestPath":"/svc/usersettings/pubcalculatorsettings","SpanId":"|d876fcf-4b964af8013a80e2.","TraceId":"d876fcf-4b964af8013a80e2","ParentId":"","ConnectionId":"0HM7JQPP7L2SV"}}
{"Timestamp":"2021-05-05T01:46:23.8480415+00:00","Level":"Debug","MessageTemplate":"{Self} Responding to {Sender} with message {Message}","Properties":{"Self":"192.168.108.208:44300/partition-identity","Sender":"192.168.108.208:44300/$1313441","Message":"{ \"pid\": 192.168.108.208:44300/partition-activator/calculatorsettings/pubsettings$5 }","SourceContext":"Proto.Context.ActorContext","ActionId":"583707b7-28d4-4cb6-8325-91b582d0dd88","ActionName":"PVL.UserSettings.Svc.Controllers.PubCalculatorSettingsController.Get (PVL.UserSettings.Svc)","RequestId":"0HM7JQPP7L2SV:00000001","RequestPath":"/svc/usersettings/pubcalculatorsettings","SpanId":"|d876fcf-4b964af8013a80e2.","TraceId":"d876fcf-4b964af8013a80e2","ParentId":"","ConnectionId":"0HM7JQPP7L2SV"}}`

**Comment by rogeralsing** [comment]
Any information on what flavor of Kubernetes you are hosting this in?
e.g. Azure Kubernetes services or just a local standard version?

Does it return null consistently, or does it eventually correct itself?

If you check the PidCache, does that contain anything for this actor?
e.g. cluster.PidCache.TryGet(clusterIdentity)


>I can see that it has identified which cluster member owns the identity:

I think this is probably the issue:
```
{
    "Timestamp": "2021-05-05T01:36:09.9440544+00:00",
    "Level": "Debug",
    "MessageTemplate": "Identity belongs to {address}",
    "Properties": {
        "address": "",    //<---- empty
```

^ "address" property is empty there, so something seems off.

If there are no members that can activate a specific type, it should retry and output that no member for kind was found.
But this looks like it got past that bit, but for whatever reason has empty member information for some reason.

**Comment by sudsy** [comment]
I finally got around to looking at this again.

For some reason there seems to be some connectivity issues when using the Kubernetes provider. See the log below:

[usersettings-log.txt](https://github.com/AsynkronIT/protoactor-dotnet/files/6595920/usersettings-log.txt)

Early on in the process there is an EndpointTerminated event.

If I switch back to the consul provider this problem disappears. I tried lowering the bar for connecting the members by removing the authentication between them but that had no impact. 

I'm running inside an AWS EKS cluster running K8s 1.18

Perhaps, this line sheds some light on what's going on - 

`[Cluster] Unable to watch the cluster status: {Error}","Properties":{"Error":"An element with the same key but a different value already exists. Key: '192.168.120.220:44300'`
